File: lru-cache-project/src/lru_cache.py
from .linked_list import DoublyLinkedList, Node


class LRUCache:
    def __init__(self, capacity):
        if capacity <= 0:
            raise ValueError("capacity must be a positive integer")
        self.capacity = capacity
        self._map = {}
        self._list = DoublyLinkedList()
        self.hits = 0
        self.misses = 0

    def __len__(self):
        return len(self._map)

    def get(self, key):
        if key not in self._map:
            self.misses += 1
            return -1
        node = self._map[key]
        self._list.move_to_head(node)
        self.hits += 1
        return node.value

    def put(self, key, value):
        if key in self._map:
            node = self._map[key]
            node.value = value
            self._list.move_to_head(node)
            return

        node = Node(key, value)
        self._map[key] = node
        self._list.add_to_head(node)

        if len(self._map) > self.capacity:
            lru_node = self._list.remove_tail()
            del self._map[lru_node.key]

    def hit_rate(self):
        """Return the fraction of get() calls that were hits, as a float 0.0-1.0.
        Returns 0.0 if get() has never been called (avoids division by zero)."""
        total = self.hits + self.misses
        if total == 0:
            return 0.0
        return self.hits / total

File: lru-cache-project/java/LRUCache.java
import java.util.HashMap;

public class LRUCache {
    private int capacity;
    private HashMap<Integer, Node> map;
    private DoublyLinkedList list;
    private int hits;
    private int misses;

    public LRUCache(int capacity) {
        if (capacity <= 0) {
            throw new IllegalArgumentException("capacity must be a positive integer");
        }
        this.capacity = capacity;
        this.map = new HashMap<>();
        this.list = new DoublyLinkedList();
        this.hits = 0;
        this.misses = 0;
    }

    public int size() {
        return map.size();
    }

    public int get(int key) {
        if (!map.containsKey(key)) {
            misses++;
            return -1;
        }
        Node node = map.get(key);
        list.moveToHead(node);
        hits++;
        return node.value;
    }

    public void put(int key, int value) {
        if (map.containsKey(key)) {
            Node node = map.get(key);
            node.value = value;
            list.moveToHead(node);
            return;
        }

        Node node = new Node(key, value);
        map.put(key, node);
        list.addToHead(node);

        if (map.size() > capacity) {
            Node lruNode = list.removeTail();
            map.remove(lruNode.key);
        }
    }

    public double hitRate() {
        int total = hits + misses;
        if (total == 0) {
            return 0.0;
        }
        return (double) hits / total;
    }
}